### PR TITLE
APM DBM Link for Postgres

### PR DESCRIPTION
**Pull request by crysmags**
### What does this PR do?

This PR is to enable support for SQL comment injection when set to service. 

### Plugin Checklist

- [x] Unit tests.
- [x] TypeScript [definitions][1].
- [x] TypeScript [tests][2].
- [x] API [documentation][3].
- [ ] CircleCI [jobs/workflows][4].
- [ ] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/dd-trace-js/blob/master/docs/API.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### CODE REVIEW

File: packages/dd-trace/src/config.js
Comment (bengl) [inline packages/dd-trace/src/config.js:83]: "sql injection" has a lot of connotations that we'd best avoid here. Since this is specifically for DBM, maybe we can call this `dmInjectionMode`?

File: packages/dd-trace/src/plugins/database.js
Comment (bengl) [inline packages/dd-trace/src/plugins/database.js:8]: Again, let's try to avoid calling this "sql injection". Maybe `createDBMComment`?

**Review by bengl — CHANGES_REQUESTED** [review]

File: packages/dd-trace/src/config.js
Comment (Qard) [reply packages/dd-trace/src/config.js:83]: This is a cross-language feature, is it not? Have we not had a discussion about what the config names should be across languages? If not, we probably _should_ before landing this.

File: packages/dd-trace/src/plugins/database.js
Comment (tlhunter) [reply packages/dd-trace/src/plugins/database.js:8]: I think this method would look a little cleaner with this syntax:

```js
    createDBMComment() {
      //  body
    }
```

Comment (crysmags) [reply packages/dd-trace/src/plugins/database.js]: updated this to
` createDBMPropagationComment() {
}`

**Comment by crysmags** [comment]
I am still failing the lint test because the length of one my tests includes the whole comment and increase the length past the limit.
```
Error:   284:1  error  This line has a length of 145. Maximum allowed is 120  max-len
Error:   334:1  error  This line has a length of 185. Maximum allowed is 120  max-len
```

File: packages/dd-trace/src/config.js
Comment (crysmags) [reply packages/dd-trace/src/config.js]: this has changed to `dbmPropagationMode` and `DD_DBM_PROPAGATION_MODE` according to the [doc](https://docs.google.com/document/d/1hlp3CuIRE-os2zYgc0wYOS8kPtCKqlrGL7aepbHHDaU/edit#).

File: packages/datadog-instrumentations/src/pg.js
Comment (tlhunter) [inline packages/datadog-instrumentations/src/pg.js:45]: @bengl, did you have concerns about the name here? I remember you saying something about not wanting to tie the published message payload with the library variable name?

File: packages/datadog-plugin-pg/src/index.js
Comment (tlhunter) [inline packages/datadog-plugin-pg/src/index.js:14]: @bengl do we have a one-time-warning helper or anything?

A future value that customers will be able to set is called `all`, and I think if they provide that value we might want to print a warning then fallback to `service`.

Comment (tlhunter) [reply packages/datadog-plugin-pg/src/index.js:14]: Though, on second thought, such a call should really happen here:
https://github.com/DataDog/dd-trace-js/pull/2480/files#diff-35b625667f1fbe3cfe5d2d1290484d9643c044d90ece6dea383d9e26a1633755R82

**Review by tlhunter — DISMISSED** [review]
This LGTM but we'll need another approver.

**Review by bengl — CHANGES_REQUESTED** [review]

File: packages/datadog-instrumentations/src/pg.js
Comment (rochdev) [reply packages/datadog-instrumentations/src/pg.js:45]: It's a big redundant since it's published by `pg` in the first place. Just `query` might make more sense.

File: packages/datadog-plugin-pg/src/index.js
Comment (rochdev) [reply packages/datadog-plugin-pg/src/index.js:14]: Should this condition be in every database plugin? Maybe there could be an additional helper in the base class, something like `pgQuery.text = this.injectDBM(pgQuery.text)` so that when it's disabled it simply returns the string as-is, and then the condition and where to add the comment can be centralized.

**Review by rochdev — CHANGES_REQUESTED** [review]

File: packages/datadog-instrumentations/src/pg.js
Comment (bengl) [reply packages/datadog-instrumentations/src/pg.js:45]: > Just `query` might make more sense.

Comment (crysmags) [reply packages/datadog-instrumentations/src/pg.js:45]: 👍

File: docs/API.md
Comment (bengl) [reply docs/API.md:390]: Indeed, the spec here says `disabled`.

Comment (tlhunter) [reply docs/API.md:390]: I like the internal representation of disabled as `false` to simplify if statements. But yeah the input space is ultimately an enum of `service`, `full`, and `disabled`.

Comment (crysmags) [reply docs/API.md:390]: @bengl advice here

Comment (crysmags) [reply docs/API.md:390]: guess I should have refreshed ha. So are we changing false to an empty string?

Comment (crysmags) [reply docs/API.md:390]: or should we use `disabled`

File: packages/datadog-instrumentations/src/pg.js
Comment (crysmags) [reply packages/datadog-instrumentations/src/pg.js:45]: @bengl and @rochdev `pgQuery` is actually coming from  `const pgQuery = queryQueue[queryQueue.length - 1] || activeQuery`

Comment (crysmags) [reply packages/datadog-instrumentations/src/pg.js:45]: would you like me to change both to query or keep the same?

File: packages/datadog-plugin-pg/src/index.js
Comment (crysmags) [reply packages/datadog-plugin-pg/src/index.js]: this has been updated to push the logic up to the db class

File: docs/API.md
Comment (rochdev) [reply docs/API.md:390]: If the spec says `disabled` let's go with that, especially since it makes it clearer what the behaviour actually is compared to an empty string.

Comment (crysmags) [reply docs/API.md]: The docs and the config have been updated to default to `disabled`

File: packages/datadog-instrumentations/src/pg.js
Comment (crysmags) [reply packages/datadog-instrumentations/src/pg.js]: pgQuery has been updated to query in the params as well as in the pg plugin index

**Review by Qard — APPROVED** [review]

File: packages/dd-trace/src/plugins/database.js
Comment (tlhunter) [inline packages/dd-trace/src/plugins/database.js:30]: Note that `carrier.traceparent` is generated by us and is not at risk of having special characters which is why it bypasses `encodeURIComponent()`.

File: packages/dd-trace/src/config.js
Comment (Qard) [inline packages/dd-trace/src/config.js:90]: Actually, looking closer, this will happen before the logger is configured and thus impossible to ever see. I'm making a related change here which could perhaps be pulled out to a separate PR: https://github.com/DataDog/dd-trace-js/pull/2477/commits/3b4da808e753b3c60c172a3f480b27b19b5c320a#diff-35b625667f1fbe3cfe5d2d1290484d9643c044d90ece6dea383d9e26a1633755R63-R75

**Review by Qard — CHANGES_REQUESTED** [review]

File: packages/dd-trace/src/config.js
Comment (Qard) [reply packages/dd-trace/src/config.js:90]: Opened a PR for that: https://github.com/DataDog/dd-trace-js/pull/2505

Comment (rochdev) [reply packages/dd-trace/src/config.js:90]: Actually, isn't it supported now anyway?

File: packages/dd-trace/src/plugins/database.js
Comment (rochdev) [inline packages/dd-trace/src/plugins/database.js:25]: It can be unsafe to use a file variable to store information, especially if it will be injected in customer data. It would also be better in any case to extract the logic from `_injectTraceparent` to a separate helper that would be used by both `TextMapPropagator` and DBM instead of using a private method with an unnecessary carrier.

File: packages/dd-trace/src/plugins/database.js
Comment (rochdev) [inline packages/dd-trace/src/plugins/database.js:14]: These variables will be re-encoded on every query, but their values will never change for the lifetime of the process so their values should be cached to avoid the cost of encoding each variable each time.

Comment (rochdev) [reply packages/dd-trace/src/plugins/database.js:14]: Or alternatively the resulting string including all variables could be cached but that would be more difficult to implement. Either way should be fast enough.

**Review by rochdev — CHANGES_REQUESTED** [review]

File: packages/dd-trace/src/config.js
Comment (Qard) [reply packages/dd-trace/src/config.js:90]: Yeah, seems like the error is out of date.

Comment (tlhunter) [reply packages/dd-trace/src/config.js:90]: Good point. This whole `if` statement should be removed.

File: packages/dd-trace/src/plugins/database.js
Comment (tlhunter) [reply packages/dd-trace/src/plugins/database.js:14]: We can throw this at the top:

```js
let dbmComment = ''
```

And then in the fuction:

```js
if (dbmComment) return dbmComment
...
dbmComment = `...`
return dbmComment
```

Or, to remove the slight perf hit of an `if` statement at runtime, should we instead do a method swap?

Comment (rochdev) [reply packages/dd-trace/src/plugins/database.js:14]: Dynamic config will allow changing the service name at runtime, so a simple cache without any checks wouldn't really be enough unfortunately. We would have to first deprecate setting these options programmatically. This is why I mentioned it would be more difficult to implement than caching the values directly.

File: packages/dd-trace/src/config.js
Comment (crysmags) [reply packages/dd-trace/src/config.js:90]: very true removing the if block

File: docs/API.md
Comment (crysmags) [reply docs/API.md]: yes, I have update to include full propagation

File: packages/dd-trace/src/plugins/database.js
Comment (tlhunter) [reply packages/dd-trace/src/plugins/database.js:14]: Is the service name the only part that can dynamically change at runtime?

Comment (rochdev) [reply packages/dd-trace/src/plugins/database.js:14]: No it was just an example but all of the values can.

Comment (tlhunter) [reply packages/dd-trace/src/plugins/database.js:14]: In that case we should be able to compare the four input values (`this.config.service` and the three `this.tracer._*` values) against a locally cached version. If the values differ then recalculate the return value, cache it, and return it. Later, if the four values are equal, then return the cached return value.

File: packages/dd-trace/src/plugins/database.js
Comment (crysmags) [reply packages/dd-trace/src/plugins/database.js:25]: @rochdev what do you mean by extract logic? Rebuild the logic in a helper or use the function within a helper to pull out what I need?

File: packages/dd-trace/src/plugins/database.js
Comment (crysmags) [reply packages/dd-trace/src/plugins/database.js]: I have updated this to only encode on the instance that the tags have changed

File: packages/dd-trace/src/plugins/database.js
Comment (tlhunter) [reply packages/dd-trace/src/plugins/database.js]: @rochdev can we throw a static method on the `TextMapPropagator` class and call it a helper? Or do you think it would better belong in another location?

```js
static getTraceparent(spanContext) {
  const sampling = spanContext._sampling.priority >= AUTO_KEEP ? '01' : '00'
  const traceId = spanContext._traceId.toString(16).padStart(32, '0')
  const spanId = spanContext._spanId.toString(16).padStart(16, '0')
  return `01-${traceId}-${spanId}-${sampling}`
}
```

Comment (rochdev) [reply packages/dd-trace/src/plugins/database.js]: @crysmags I meant to move the logic to a helper than is then imported and used by both `TextMapPropagator` and the plugin class.

@tlhunter I think it should probably go in `datadog-core` if it will be used in both plugin and dd-trace, especially with IDs and W3C tracecontext being core concepts for an APM, but I don't have a strong opinion since we can always move it later, but it shouldn't require exposing an internal class of the tracer.

File: packages/dd-trace/src/plugins/database.js
Comment (rochdev) [inline packages/dd-trace/src/plugins/database.js:28]: This is the same code repeated 4 times with only 2 property names changing, and thus could be extracted to a function called 4 times instead of repeating the code.

File: packages/dd-trace/src/plugins/database.js
Comment (rochdev) [inline packages/dd-trace/src/plugins/database.js:54]: The config should never be changed from a plugin.

File: packages/dd-trace/src/plugins/database.js
Comment (rochdev) [inline packages/dd-trace/src/plugins/database.js:55]: This is repeated in both branches of the condition.

File: packages/dd-trace/src/plugins/database.js
Comment (rochdev) [inline packages/dd-trace/src/plugins/database.js:45]: This line could be smaller by using additional variables.

**Review by rochdev — CHANGES_REQUESTED** [review]

File: packages/dd-trace/src/plugins/database.js
Comment (rochdev) [reply packages/dd-trace/src/plugins/database.js]: The important part is not really where it is, but that it should be private and it shouldn't require an object to be passed since that's unique to `TextMapPropagator`.

Comment (crysmags) [reply packages/dd-trace/src/plugins/database.js]: Ok I created a helper in the db plugin in order to not use object and wrote a to do to create an additional helper that could be used by both text map and the plugins

File: packages/dd-trace/src/plugins/database.js
Comment (crysmags) [reply packages/dd-trace/src/plugins/database.js:59]: @rochdev I discussed this with Thomas as well, I put them under the conditionals of this functions because leaving at the top means it is called regardless if propagation is enabled or not. Currently the function (`injectDbmQuery`) is called in the plugin without any conditions needing to be met.

File: packages/dd-trace/src/plugins/database.js
Comment (crysmags) [reply packages/dd-trace/src/plugins/database.js:23]: created a helper called `propagationHelper` which will take in the tags and update the local service tags object

File: packages/dd-trace/src/plugins/database.js
Comment (crysmags) [reply packages/dd-trace/src/plugins/database.js]: I removed this since I am no longer calling the text map function.

File: packages/dd-trace/src/plugins/database.js
Comment (crysmags) [reply packages/dd-trace/src/plugins/database.js:55]: @tlhunter @rochdev @bengl anything to update here?

Comment (bengl) [reply packages/dd-trace/src/plugins/database.js:55]: @crysmags One way to handle this is to early-return if propagation is disabled. This default is `none` right? so you could do something like:

```js
const mode = this.config.dbmPropagationMode
if (mode === 'none') {
  return query
}
const servicePropagation = this.createDBMPropagationCommentService()
if (mode === 'service') {
  return `/*${servicePropagation}*/ ${query}`
} else if (mode === 'full') {
  config.experimental.traceparent = true
  textMap._injectTraceparent(this.activeSpan._spanContext, carrier)
  return `/*${servicePropagation},traceparent='${carrier.traceparent}'*/ ${query}`
} else {
  // This shouldn't happen because mode can only be 'none', 'service', or 'full', but
  // in case it's something else, we'll return the query
  return query
}

Comment (crysmags) [reply packages/dd-trace/src/plugins/database.js]: got it, updated this function

File: packages/dd-trace/src/plugins/database.js
Comment (rochdev) [inline packages/dd-trace/src/plugins/database.js:21]: This could have a more descriptive name, and should probably also be a verb as it's a method.

File: packages/dd-trace/src/plugins/database.js
Comment (rochdev) [inline packages/dd-trace/src/plugins/database.js:29]: This variable name is not very descriptive. However, it's actually not even needed and instead individual variables could be used with for example `const { encodedDddbd  } = this.serviceTags`.

File: packages/dd-trace/src/plugins/database.js
Comment (rochdev) [inline packages/dd-trace/src/plugins/database.js:39]: Since it's just a matter of putting this function elsewhere, let's do it in this PR to avoid code duplication.

**Review by rochdev — CHANGES_REQUESTED** [review]
Just a few nitpicks, but otherwise everything LGTM.

File: packages/dd-trace/src/plugins/database.js
Comment (crysmags) [reply packages/dd-trace/src/plugins/database.js:21]: updated `propagationHelper` to `encodingServiceTags`

File: packages/dd-trace/src/plugins/database.js
Comment (crysmags) [reply packages/dd-trace/src/plugins/database.js:29]: Very true, updated to `const { encodedDddbs, encodedDde, encodedDdps, encodedDdpv } = this.serviceTags` which is now called after using the `encodingServiceTags` helper

File: packages/dd-trace/src/plugins/database.js
Comment (crysmags) [reply packages/dd-trace/src/plugins/database.js:39]: Sure thing! Still not sure where @rochdev @bengl @Qard @tlhunter let me know where I can add this function. Thomas suggested previously creating a file in `datadog-core`, such as
`packages/datadog-core/src/traceparent.js`. Let me know if this is a good spot.

Comment (rochdev) [reply packages/dd-trace/src/plugins/database.js:40]: Yeah that's fine. It can always be moved later if we need, I just want to avoid the duplicated code.

Comment (Qard) [reply packages/dd-trace/src/plugins/database.js:40]: Maybe we could put it on SpanContext as `toTraceparent`, similar to how we currently have `toTraceId` and `toSpanId`?

**Review by Qard — APPROVED** [review]
LGTM with one small question.

File: packages/dd-trace/src/plugins/database.js
Comment (rochdev) [reply packages/dd-trace/src/plugins/database.js:40]: It's worth noting that these 2 methods are from OpenTracing and if we add methods we're deviating from the spec. This type of change will be easier when the API is decoupled from actual implementation.

**Review by rochdev — APPROVED** [review]

**Pull request merged**
